pyvrp/cpp/search/LocalSearch.cpp

```diff
@@ -251,47 +254,110 @@ void LocalSearch::applyEmptyRouteMoves(Route::Node *U,
 {
-    auto const uClient = U->client();
-    ProblemData::Client const &uData = data.location(uClient);
+    ProblemData::Client const &uData = data.location(U->client());
+
+    auto mustInsert = uData.required;
+    if (uData.group)
+    {
+        auto const &group = data.group(uData.group.value());
+        auto const pred = [&](auto client) { return nodes[client].route(); };
+        auto const groupInSol = std::any_of(group.begin(), group.end(), pred);
+
+        if (groupInSol)  // then we'll have to wait for the group moves to
+            return;      // determine what to do.
+
+        // U must be inserted when it is in a required mutually exclusive client
+        // group that currently is not in the solution at all.
+        assert(group.mutuallyExclusive);
+        mustInsert = group.required && !groupInSol;
+    }
 
-    // First test removing U. This is allowed when U is not required.
     if (!uData.required && removeCost(U, data, costEvaluator) < 0)
     {
         auto *route = U->route();
         route->remove(U->idx());
         update(route, route);
     }
 
-    // If U is not currently in the solution, we test if inserting it is an
-    // improving move. Note that we always insert required clients.
-    if (!U->route())
-    {
-        // We take this as a default value in case none of the client's
-        // neighbours are assigned, yet U is required.
-        Route::Node *UAfter = routes[0][0];
-        Cost bestCost = insertCost(U, UAfter, data, costEvaluator);
+    if (U->route())  // then we are done, since having U is better than
+        return;      // removing it.
 
-        for (auto const vClient : neighbours_[uClient])
-        {
-            auto *V = &nodes[vClient];
+    // We take this as a default value in case none of the client's neighbours
+    // are assigned, yet U is required.
+    Route::Node *UAfter = routes[0][0];
+    Cost bestCost = insertCost(U, UAfter, data, costEvaluator);
 
-            if (!V->route())
-                continue;
+    for (auto const vClient : neighbours_[U->client()])
+    {
+        auto *V = &nodes[vClient];
 
-            auto cost = insertCost(U, V, data, costEvaluator);
-            if (cost < bestCost)
-            {
-                bestCost = cost;
-                UAfter = V;
-            }
-        }
+        if (!V->route())
+            continue;
 
-        if (uData.required || bestCost < 0)
+        auto cost = insertCost(U, V, data, costEvaluator);
+        if (cost < bestCost)
         {
-            UAfter->route()->insert(UAfter->idx() + 1, U);
-            update(UAfter->route(), UAfter->route());
+            bestCost = cost;
+            UAfter = V;
         }
     }
+
+    if (mustInsert || bestCost < 0)
+    {
+        UAfter->route()->insert(UAfter->idx() + 1, U);
+        update(UAfter->route(), UAfter->route());
+    }
+}
+
+void LocalSearch::applyGroupMoves(Route::Node *U,
```

Comment:
Just to summarize for myself:
- Assumption: if U has a group, then at least one client of the group must be in the solution.
	- This assumption is satisfied, because `applyOptionalClientMoves` makes sure that if a required group has no clients in the solution, then it will be inserted.
- Find all inserted group clients and remove all but the one with the lowest delta cost. This is V.
- Swap U for V if U != V and if improving.

Reply:
Si!